protocols/gossipsub/src/handler.rs

```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
             Self::OutboundOpenInfo,
         >,
     ) {
+        if event.is_outbound() {
+            self.outbound_substream_establishing = false;
+        }
+
+        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many inbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of inbound substreams created has been exceeded.");
+            return;
+        }
+
+        if event.is_outbound() && self.outbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many outbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of outbound substreams created has been exceeded.");
+            return;
+        }
+
         match event {
             ConnectionEvent::FullyNegotiatedInbound(fully_negotiated_inbound) => {
                 self.on_fully_negotiated_inbound(fully_negotiated_inbound)
             }
             ConnectionEvent::FullyNegotiatedOutbound(fully_negotiated_outbound) => {
                 self.on_fully_negotiated_outbound(fully_negotiated_outbound)
             }
-            ConnectionEvent::DialUpgradeError(DialUpgradeError { error: e, .. }) => {
-                self.outbound_substream_establishing = false;
-                warn!("Dial upgrade error {:?}", e);
-                self.upgrade_errors.push_back(e);
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer,
+                ..
+            }) => {
+                log::debug!("Dial upgrade error: Protocol negotiation timeout.");
```

Comment:
We only do the negotiation once to create long-lived substreams. If the negotiation times out, I rarely find it useful to continually re-connect and retry the negotiations, which is why I previously opted to just close the connection. 

If I understand these changes, every message we want to send, we are going to continually try and re-negotiate?

Reply:
Currently, we'd try `MAX_SUBSTREAM_CREATION` times and then disable the handler.

Reply:
Yep, that seems like reasonable logic. 

If there were another behaviour that had a keepalive::Yes. It would continue to try and reconnect tho right?

Reply:
> Currently, we'd try `MAX_SUBSTREAM_CREATION` times and then disable the handler.

I don't think this is correct. We increase `outbound_substream_established` only on successful substream negotiation (i.e. in `on_fully_negotiated_outbound`). The above failure does not increase the counter and thus the limit `MAX_SUBSTREAM_CREATION` is never hit.

Reply:
> > Currently, we'd try `MAX_SUBSTREAM_CREATION` times and then disable the handler.
> 
> I don't think this is correct. We increase `outbound_substream_established` only on successful substream negotiation (i.e. in `on_fully_negotiated_outbound`). The above failure does not increase the counter and thus the limit `MAX_SUBSTREAM_CREATION` is never hit.

You are totally right, not sure how I missed that. Do we agree that the logic I described is what we want to have?

Reply:
> Do we agree that the logic I described is what we want to have?

Agreed in case it includes the re-enqueuing of the RPC of the failed outbound stream. See https://github.com/libp2p/rust-libp2p/pull/3625#discussion_r1150749684.

Reply:
Tracking the number of requested outbound streams instead of the number of successfully upgraded outbound streams is done with https://github.com/libp2p/rust-libp2p/pull/3625/commits/0507493a1e5232047c8495319e82b00bd80303eb.